compiler/src/lnntoamm/Types.ts

```diff
@@ -123,127 +193,279 @@ export default abstract class Type implements Equalable {
   }
 }
 
-class Builtin extends Type {
+class Opaque extends Type implements Generalizable {
+  // if any values are `null` that means that this isn't an instantiable type
+  // and should be treated like the type needs to be duped
+  generics: GenericArgs;
+
   get ammName(): string {
-    return this.name;
+    let generics = '';
+    if (Object.keys(this.generics).length !== 0) {
+      const genNames = new Array<string>();
+      for (const [tyVar, ty] of Object.entries(this.generics)) {
+        if (ty === null) {
+          genNames.push(tyVar);
+        } else {
+          genNames.push(ty.ammName);
+        }
+      }
+      generics = '<' + genNames.join(', ') + '>';
+    }
+    return this.name + generics;
   }
 
-  constructor(name: string) {
+  constructor(name: string, generics: string[]) {
     super(name);
+    this.generics = {};
+    generics.forEach((g) => (this.generics[g] = null));
   }
 
-  compatibleWithConstraint(ty: Type, scope: Scope): boolean {
-    if (ty instanceof Builtin) {
-      return this.eq(ty);
-    } else if (ty instanceof OneOf || ty instanceof Generated) {
-      return ty.compatibleWithConstraint(this, scope);
-    } else if (ty instanceof HasOperator) {
-      return Has.operator(ty, scope, this).length !== 0;
-    } else if (ty instanceof HasMethod) {
-      return Has.method(ty, scope, this).length !== 0;
-    } else if (ty instanceof HasField) {
-      return TODO('add field support for builtins');
-    } else if (ty instanceof Interface) {
-      // FIXME: once builtins can have fields
+  compatibleWithConstraint(that: Type, scope: Scope): boolean {
+    if (this.eq(that)) {
+      return true;
+    }
+    if (that instanceof Opaque) {
+      const thisGens = Object.values(this.generics);
+      const thatGens = Object.values(that.generics);
+      if (this.name !== that.name || thisGens.length !== thatGens.length) {
```

Comment:
Is the first half of this conditional true? Wouldn't `Array<any>` be compatible with `Array<int64>` but not have the same name?

Reply:
the name would be `Array` in both cases, whereas the `ammName` would be `Array<any>` or `Array<int64>`

Reply:
Got it. I called that `baseName` and your `ammName` would be the `typename`.